python/ray/util/collective/collective_group/nccl_collective_group.py

```diff
@@ -286,23 +287,142 @@ def reducescatter(self,
+        comm.send(ptr, n_elems, dtype, peer_p2p_rank, stream.ptr)
+
+    def recv(self, tensor, src_rank):
+        """Receive tensor from a source process in the group.
+
+        Args:
+            tensor: the received tensor.
+            src_rank: the rank of the source process.
+
+        Returns:
+            None
+        """
+        if nccl_util.get_nccl_runtime_version() < 2704:
+            raise RuntimeError("recv is not available requires NCCL >= 2.7.4. "
+                               "Got '{}'.".format(
+                                   nccl_util.get_nccl_runtime_version()))
+        peer_p2p_rank = 0 if self.rank > src_rank else 1
+        comm = self._get_nccl_p2p_communicator(src_rank, self.rank)
+        stream = self._get_cuda_stream()
 
+        dtype = nccl_util.get_nccl_tensor_dtype(tensor)
+        ptr = nccl_util.get_tensor_ptr(tensor)
+        n_elems = nccl_util.get_tensor_n_elements(tensor)
+        comm.recv(ptr, n_elems, dtype, peer_p2p_rank, stream.ptr)
+
+    def _get_nccl_collective_communicator(self):
+        """Create or retrieve a cached NCCL communicator.
+
+        Returns:
+            communicator
         """
-        # TODO(Hao): later change this to use device keys and query from cache.
         # TODO(Hao): implement a thin wrapper
-        if not self._nccl_comm:
-            self._nccl_comm = nccl_util.create_nccl_communicator(
-                self.world_size, self.nccl_uid, self.rank)
-        return self._nccl_comm
+        if not self._collective_comm_cache:
+            # create the communicator
+            if self.rank == 0:
+                group_uid = self._generate_nccl_uid(self.group_name)
+            else:
+                rendezvous = Rendezvous(self.group_name)
+                rendezvous.meet()
+                group_uid = rendezvous.get_nccl_id()
+            self._collective_comm_cache = \
+                nccl_util.create_nccl_communicator(self.world_size,
+                                                   group_uid,
+                                                   self.rank)
+        return self._collective_comm_cache
+
+    def _get_nccl_p2p_communicator(self, src_rank, dst_rank):
+        """Create or retrieve an NCCL communicator for p2p tasks.
+
+        Args:
+            src_rank (int): source rank.
+            dst_rank (int): destination rank.
+
+        Returns:
+            communicator
+        """
+        min_rank = min(src_rank, dst_rank)
+        max_rank = max(src_rank, dst_rank)
+        my_rank = 0 if self.rank == min_rank else 1
+        p2p_group_key = self._generate_p2p_group_key(min_rank, max_rank)
+        comm = self._p2p_comm_cache.get(p2p_group_key)
+        if not comm:
+            if self.rank == min_rank:
+                group_uid = self._generate_nccl_uid(p2p_group_key)
+            else:
+                rendezvous = Rendezvous(p2p_group_key)
+                rendezvous.meet()
+                group_uid = rendezvous.get_nccl_id()
```

Comment:
do you need to create a rendezvous for rank == min_rank too?

Reply:
we don't have to; because the rendezvous is used for other nodes to obtain the `NCCLUniqueID`, which was initialized at the node with min_rank.